DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/EndToEndTest.java

```diff
@@ -97,24 +122,57 @@ private void migrationDocumentsWithClusters(
             var sourceRepo = new FileSystemRepo(localDirectory.toPath());
 
             // === ACTION: Migrate the documents ===
+            var runCounter = new AtomicInteger();
             final var clockJitter = new Random(1);
-            var result = migrateDocumentsWithOneWorker(
-                sourceRepo,
-                snapshotName,
-                List.of(),
-                targetCluster.getUrl(),
-                clockJitter,
-                testDocMigrationContext,
-                sourceCluster.getContainerVersion().getVersion(),
-                false
+
+            // ExpectedMigrationWorkTerminationException is thrown on completion.
+            var expectedTerminationException = Assertions.assertThrows(
+                ExpectedMigrationWorkTerminationException.class,
+                () -> migrateDocumentsSequentially(
+                    sourceRepo,
+                    snapshotName,
+                    List.of(),
+                    targetCluster.getUrl(),
+                    runCounter,
+                    clockJitter,
+                    testDocMigrationContext,
+                    sourceCluster.getContainerVersion().getVersion(),
+                    false
+                )
             );
-            assertThat(result, equalTo(DocumentsRunner.CompletionStatus.WORK_COMPLETED));
+
+            Assertions.assertEquals(numberOfShards + 1, expectedTerminationException.numRuns);
 
             // Check that the docs were migrated
             checkClusterMigrationOnFinished(sourceCluster, targetCluster, testDocMigrationContext);
+
+            // Check that that docs were migrated with routing
+            checkDocsWithRouting(sourceCluster, testDocMigrationContext);
+            checkDocsWithRouting(targetCluster, testDocMigrationContext);
         } finally {
             deleteTree(localDirectory.toPath());
         }
     }
 
+    private void checkDocsWithRouting(
```

Comment:
Thanks for adding this function, though we are missing any assertion on the doc with routing on shard 2.

Can we take in indexName, queryString, and _routing as parameters to this function and return the hits?

Then we can do the assertion in the test above.

This function can also be moved to SourceTestBase